src/Analyzer/Passes/FunctionToSubcolumnsPass.cpp

```diff
@@ -16,218 +17,428 @@
+
+    auto column_source = first_argument_column_node->getColumnSource();
+    auto * table_node = column_source->as<TableNode>();
+    if (!table_node)
+        return {};
+
+    const auto & storage = table_node->getStorage();
+    const auto & storage_snapshot = table_node->getStorageSnapshot();
+    auto column = first_argument_column_node->getColumn();
+
+    if (!storage->supportsOptimizationToSubcolumns() || storage->isVirtualColumn(column.name, storage_snapshot->metadata))
+        return {};
+
+    auto column_in_table = storage_snapshot->tryGetColumn(GetColumnsOptions::All, column.name);
+    if (!column_in_table || !column_in_table->type->equals(*column.type))
+        return {};
+
+    return std::make_tuple(function_node, first_argument_column_node, table_node);
+}
+
+/// First pass collects info about identifiers to determine which identifiers are allowed to optimize.
+class FunctionToSubcolumnsVisitorFirstPass : public InDepthQueryTreeVisitorWithContext<FunctionToSubcolumnsVisitorFirstPass>
 {
 public:
-    using Base = InDepthQueryTreeVisitorWithContext<FunctionToSubcolumnsVisitor>;
+    using Base = InDepthQueryTreeVisitorWithContext<FunctionToSubcolumnsVisitorFirstPass>;
     using Base::Base;
 
-    void enterImpl(QueryTreeNodePtr & node) const
+    void enterImpl(const QueryTreeNodePtr & node)
     {
         if (!getSettings().optimize_functions_to_subcolumns)
             return;
 
-        auto * function_node = node->as<FunctionNode>();
-        if (!function_node)
+        if (auto * table_node = node->as<TableNode>())
+        {
+            enterImpl(*table_node);
             return;
+        }
 
-        auto & function_arguments_nodes = function_node->getArguments().getNodes();
-        size_t function_arguments_nodes_size = function_arguments_nodes.size();
-
-        if (function_arguments_nodes.empty() || function_arguments_nodes_size > 2)
+        if (auto * column_node = node->as<ColumnNode>())
+        {
+            enterImpl(*column_node);
             return;
+        }
 
-        auto * first_argument_column_node = function_arguments_nodes.front()->as<ColumnNode>();
-
-        if (!first_argument_column_node)
+        auto [function_node, first_argument_node, table_node] = getTypedNodesForOptimization(node);
+        if (function_node && first_argument_node && table_node)
+        {
+            enterImpl(*function_node, *first_argument_node, *table_node);
             return;
+        }
 
-        if (first_argument_column_node->getColumnName() == "__grouping_set")
+        if (const auto * join_node = node->as<JoinNode>())
+        {
+            has_join_use_nulls |= getContext()->getSettingsRef().join_use_nulls;
             return;
+        }
+    }
 
-        auto column_source = first_argument_column_node->getColumnSource();
-        auto * table_node = column_source->as<TableNode>();
+    std::unordered_set<Identifier> getIdentifiersToOptimize() const
+    {
+        if (has_join_use_nulls)
```

Comment:
What about `group_by_use_nulls`? It may have the same problem.

Reply:
Yes, optimization should be disabled. Fixed in the latest commit.